### ISSUE TITLE
lazy component not loading parent dependencies

### ISSUE THREAD

**Issue by denisyilmaz** [angular/angular#36193]
# 🐞 bug report

### Affected Package
Angular 9.0.7 and Angular 9.1.0-rc.0

### Is this a regression?
Yes, lazy components were working in the first versions of 9.0.0, now with 9.0.5-9.0.7 the bug appeared.

### Description

When loading a component lazy with the resolveComponentFactory() described in different [Tutorials](https://johnpapa.net/angular-9-lazy-loading-components/) the component don't inherit the CommonModule or any other Dependencies from the parent Module where they are loaded.

## 🔬 Minimal Reproduction

In this example I reduced it to two child components. the first is declared in app.module and is loading just fine. the second is loaded without any previous declarations and should therefore be lazy loaded.

https://stackblitz.com/edit/angular-9-0-7-lazy-component

## 🔥 Exception or Error
I tried to reproduce with stackblitz but there I get different error: 

```bash
No component factory found for Child2Component. Did you add it to @NgModule.entryComponents?
```

In my local environment this error is not appearing but I get rather the following error messages, which is the error I'm referring to:

```bash
error NG8002: Can't bind to 'ngSwitch' since it isn't a known property of 'ul'.
3   [ngSwitch]="item.switch"
```

In my project I use different pipes and directives that are shared via a SharedModule and is loaded to the parent module where the lazy components are attached. All these dependencies are not available in the lazy child components. Only empty components with no other dependencies are working.

## 🌍  Your Environment

From my project, same error was appearing in 9.0.5 and 9.0.7
**Angular Version:**
<pre><code>
Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.7
@angular-devkit/build-angular     0.901.0-next.4
@angular-devkit/build-optimizer   0.901.0-next.4
@angular-devkit/build-webpack     0.901.0-next.4
@angular-devkit/core              9.0.7
@angular-devkit/schematics        9.1.0-next.4
@angular/cdk                      9.2.0-next.0
@angular/cli                      9.1.0-next.4
@ngtools/webpack                  9.1.0-next.4
@nguniversal/builders             9.0.2
@nguniversal/common               9.0.2
@nguniversal/express-engine       9.0.2
@schematics/angular               9.1.0-next.4
@schematics/update                0.901.0-next.4
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.42.0
</code></pre>

**Comment by JoostK** [comment]
From what I can tell in the StackBlitz case, it doesn't work because `Child2Component` is not declared in any NgModule. Looking at the tutorial, this statement is _not_ accurate:

> We won't want either component to be declared in a module, since we want to lazy load them. If we declare them in a module, then Angular will eagerly load them.

Directives/components should still be declared in an NgModule as that determines their _compilation scope_, i.e. all declarations that are available to be used in the template. If you omit registration in an NgModule, no compilation scope will be available so none of e.g. `CommonModule` is available.

Regarding the "eager loading", that is not true: the compiled output will ensure there's no reference from an `NgModule` to any of its declarations, to enable for tree-shaking of declarations if they are not used.

Directives/components do not get access to other declarations (directives/components/pipes) depending on the NgModule that they are loaded in. In which NgModule a component is loaded only affects how DI works, as each individually loaded `NgModule` will have its own `Injector`. I recently commented on a similar issue with some more explanation on this in https://github.com/angular/angular/issues/35759#issuecomment-601444856.

In conclusion, there is no bug here. If there is a problem with documentation on angular.io we should look into fixing that, and perhaps @johnpapa can revisit some of the tutorial to more accurately reflect that.

**Comment by denisyilmaz** [comment]
Did I understand it correctly that it is currently not possible to load lazy components with any dependencies? Or only the way you described it in your [#35759 comment](https://github.com/angular/angular/issues/35759#issuecomment-601444856)?
Because when I declare the lazy component in a NgModule, I loose the ability to lazy load it (its part of the main.js bundle and not in its own JS file.

I tested to add the second components to the entryComponents as mentioned in the error but then the following error appears: 
```bash
Component Child2Component is not part of any NgModule or the module has not been imported into your module.
```
even with adding the parent injector the error does not disappear.

**Comment by JoostK** [comment]
The `entryComponents` error you originally experienced only applies to ViewEngine, not Ivy. Any `entryComponents` in Ivy are completely ignored, although in JIT mode the error you just mentioned is thrown if there's no corresponding declaration (this is a discrepancy with AOT, where there's no such error).

The way described in https://github.com/angular/angular/issues/35759#issuecomment-601444856 is currently the proper solution, AFAIK. It's also the exact pattern that the router uses.

I think that declaring the component in an NgModule that is part of the root module will not allow for proper code-splitting because of toolchain limitations, but I'm not sure to be honest. The declarations of an NgModule will be dropped in Terser as they are part of an IIFE marked `/* @__PURE__ */`, allowing for tree-shaking, but possibly this is restricting code-splitting. Anyway, the proper way to achieve code-splitting would be to have a separate NgModule that is not imported from the root NgModule.

**Comment by denisyilmaz** [comment]
ok, I just realized that when I have the simple import(someComponent) like @johnpapa described it and add the component to the parents declaration/entrycomponent the JIT compilation will not create a separate bundle, but AOT will create a separate bundle for its excluded from the main bundle.

this way I don't need to create a NgModule for each lazy component. It seems to work, its just inconsistent (non lazy loading in JIT but lazy loading in AOT)

**Comment by JoostK** [comment]
Optimizations done by earlier steps in the build pipeline (like Terser) can always affect how subsequent optimizations behave, such as code-splitting. So seeing code-splitting behave differently does not sound counter intuitive to me.

Still, I would always prefer to fully separate desired code-splits in the original source code, hence having a separate NgModule makes sense to me.

**Comment by alxhub** [comment]
Sounds reasonable to me.

**Comment by jelbourn** [comment]
Going to close this since it seems to be resolved. We have some plans to expand our docs around lazy-loading in the future which might help here.
